File: 2025/AHC052/a04.cpp
#include <bits/stdc++.h>
using namespace std;

// ====== 定数（問題では N=30, M=10, K=10 が固定ですが、入力からも読みます） ======

struct BitBoard {
    // 30x30 = 900 マスをビットで管理
    bitset<900> board;
    int visit_count;

    BitBoard() : board(), visit_count(0) {}

    // コピー用コンストラクタ
    BitBoard(const bitset<900>& b, int cnt) : board(b), visit_count(cnt) {}

    inline int idx(int x, int y) const { // x = 行(i), y = 列(j)
        return y * 30 + x;
    }

    inline bool has_visited(int x, int y) const {
        return board.test(idx(x, y));
    }

    inline void visit(int x, int y) {
        int id = idx(x, y);
        if (!board.test(id)) {
            board.set(id);
            visit_count++;
        }
    }
};

// 近傍で未訪問のマスを BFS で探す
pair<int,int> find_nearest_unvisited(
    const BitBoard& BB,
    int sx, int sy,
    const vector<vector<int>>& X_wall, // (N-1) x N : 上下（縦）を阻害
    const vector<vector<int>>& Y_wall  // N x (N-1) : 左右（横）を阻害
) {
    const int N = 30;
    queue<pair<int,int>> q;
    vector<vector<char>> vis(N, vector<char>(N, 0));
    q.push({sx, sy});
    vis[sx][sy] = 1;

    while (!q.empty()) {
        auto [x, y] = q.front(); q.pop();

        if (!BB.has_visited(x, y)) {
            return {x, y};
        }

        // 上
        if (x-1 >= 0 && !vis[x-1][y]) {
            if (X_wall[x-1][y] == 0) {
                vis[x-1][y] = 1;
                q.push({x-1, y});
            }
        }
        // 下
        if (x+1 < N && !vis[x+1][y]) {
            if (X_wall[x][y] == 0) {
                vis[x+1][y] = 1;
                q.push({x+1, y});
            }
        }
        // 左
        if (y-1 >= 0 && !vis[x][y-1]) {
            if (Y_wall[x][y-1] == 0) {
                vis[x][y-1] = 1;
                q.push({x, y-1});
            }
        }
        // 右
        if (y+1 < N && !vis[x][y+1]) {
            if (Y_wall[x][y] == 0) {
                vis[x][y+1] = 1;
                q.push({x, y+1});
            }
        }
    }
    return {-1, -1}; // 見つからなかった
}

// 目的地への最短路の「最初の一手」を BFS で求める
char get_next_step(
    int rx, int ry,
    int tx, int ty,
    const vector<vector<int>>& X_wall,
    const vector<vector<int>>& Y_wall
) {
    const int N = 30;
    if (rx == tx && ry == ty) return 'S';

    struct Node { int x, y; char first; };
    queue<Node> q;
    vector<vector<char>> vis(N, vector<char>(N, 0));

    q.push({rx, ry, 0});
    vis[rx][ry] = 1;

    while (!q.empty()) {
        Node cur = q.front(); q.pop();

        if (cur.x == tx && cur.y == ty) {
            return cur.first ? cur.first : 'S';
        }

        // 上へ
        if (cur.x - 1 >= 0 && !vis[cur.x - 1][cur.y]) {
            if (X_wall[cur.x - 1][cur.y] == 0) {
                vis[cur.x - 1][cur.y] = 1;
                char nd = cur.first ? cur.first : 'U';
                q.push({cur.x - 1, cur.y, nd});
            }
        }
        // 下へ
        if (cur.x + 1 < N && !vis[cur.x + 1][cur.y]) {
            if (X_wall[cur.x][cur.y] == 0) {
                vis[cur.x + 1][cur.y] = 1;
                char nd = cur.first ? cur.first : 'D';
                q.push({cur.x + 1, cur.y, nd});
            }
        }
        // 左へ
        if (cur.y - 1 >= 0 && !vis[cur.x][cur.y - 1]) {
            if (Y_wall[cur.x][cur.y - 1] == 0) {
                vis[cur.x][cur.y - 1] = 1;
                char nd = cur.first ? cur.first : 'L';
                q.push({cur.x, cur.y - 1, nd});
            }
        }
        // 右へ
        if (cur.y + 1 < N && !vis[cur.x][cur.y + 1]) {
            if (Y_wall[cur.x][cur.y] == 0) {
                vis[cur.x][cur.y + 1] = 1;
                char nd = cur.first ? cur.first : 'R';
                q.push({cur.x, cur.y + 1, nd});
            }
        }
    }
    return 'S';
}

// ロボット群を 1 回操作して状態を更新
void move_robot(
    BitBoard& BB,
    vector<pair<int,int>>& robots,
    const vector<vector<char>>& key_config, // K x M
    int key,
    const vector<vector<int>>& X_wall,
    const vector<vector<int>>& Y_wall
) {
    const int N = 30;
    const int M = (int)robots.size();

    for (int i = 0; i < M; ++i) {
        char d = key_config[key][i];
        int x = robots[i].first;
        int y = robots[i].second;
        int nx = x, ny = y;

        if (d == 'U') nx = x - 1;
        else if (d == 'D') nx = x + 1;
        else if (d == 'L') ny = y - 1;
        else if (d == 'R') ny = y + 1;
        else { // 'S'
            continue;
        }

        if (0 <= nx && nx < N && 0 <= ny && ny < N) {
            bool can_move = false;
            if (d == 'U' && X_wall[nx][ny] == 0) can_move = true;              // h_{nx,ny}
            else if (d == 'D' && X_wall[nx - 1][ny] == 0) can_move = true;     // h_{x, y}
            else if (d == 'L' && Y_wall[nx][ny] == 0) can_move = true;         // v_{x, y-1}（nyは既に-1済み）
            else if (d == 'R' && Y_wall[nx][ny - 1] == 0) can_move = true;     // v_{x, y}

            if (can_move) {
                robots[i] = {nx, ny};
                BB.visit(nx, ny);
            }
        }
    }
}

// ビームサーチで次状態を生成
struct State {
    BitBoard bb;
    vector<pair<int,int>> robots;
    vector<int> ans; // 押したボタン列（※原実装では後続で使っていないが移植して保持）
};

vector<State> generate_next_states(
    const BitBoard& BB,
    const vector<pair<int,int>>& robots,
    const vector<int>& ans,
    const vector<vector<char>>& key_config,
    const vector<vector<int>>& X_wall,
    const vector<vector<int>>& Y_wall
) {
    vector<State> res;
    for (int key = 0; key < 4; ++key) { // 0..3 のみ
        State st;
        st.bb = BitBoard(BB.board, BB.visit_count);
        st.robots = robots;
        st.ans = ans;

        move_robot(st.bb, st.robots, key_config, key, X_wall, Y_wall);
        st.ans.push_back(key);
        res.push_back(std::move(st));
    }
    return res;
}

// 方向文字 -> ボタン番号
int dir2key(char c) {
    switch (c) {
        case 'U': return 0;
        case 'D': return 1;
        case 'L': return 2;
        case 'R': return 3;
        default:  return 4; // 'S'
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // パラメータ読み込み
    int N_in, M_in, K_in;
    if (!(cin >> N_in >> M_in >> K_in)) return 0;
    const int N = 30, M = 10, K = 10; // ロジック上は 30/10/10 を想定

    vector<pair<int,int>> robots(M);
    for (int i = 0; i < M; ++i) {
        int a, b; cin >> a >> b;
        robots[i] = {a, b};
    }

    // Y_wall: N 行, 各行 (N-1) 文字
    vector<vector<int>> Y_wall(N, vector<int>(N-1, 0));
    for (int i = 0; i < N; ++i) {
        string s; cin >> s;
        for (int j = 0; j < N-1; ++j) Y_wall[i][j] = (s[j] - '0');
    }

    // X_wall: (N-1) 行, 各行 N 文字
    vector<vector<int>> X_wall(N-1, vector<int>(N, 0));
    for (int i = 0; i < N-1; ++i) {
        string s; cin >> s;
        for (int j = 0; j < N; ++j) X_wall[i][j] = (s[j] - '0');
    }

    // キーコンフィグ初期化
    vector<vector<char>> key_config(K, vector<char>(M, 'S'));
    // 0:U, 1:D, 2:L, 3:R を全ロボットに割り当て
    for (int m = 0; m < M; ++m) {
        key_config[0][m] = 'U';
        key_config[1][m] = 'D';
        key_config[2][m] = 'L';
        key_config[3][m] = 'R';
    }

    BitBoard BB;
    // 初期位置を訪問済みに
    for (auto &p : robots) BB.visit(p.first, p.second);

    // 以降は原実装のロジックを忠実に移植
    const int BEAM_WIDTH = 1000;
    const double TIME_LIMIT = 1.5; // 秒
    const int LEN = 500;

    vector<int> ans_list; // 出力する操作列

    // ビームサーチ（※原実装同様、結果は実状態に反映しない）
    vector<State> current_states;
    current_states.push_back(State{BB, robots, {}});

    std::mt19937_64 rng(1317);
    std::uniform_real_distribution<double> dist(0.0, 1.0);

    auto t0 = chrono::steady_clock::now();

    for (int step = 0; step < LEN; ++step) {
        auto t1 = chrono::steady_clock::now();
        double elapsed = chrono::duration<double>(t1 - t0).count();
        if (elapsed > TIME_LIMIT) break;

        vector<State> next_states;
        next_states.reserve(current_states.size() * 4);

        for (const auto& st : current_states) {
            auto tmp = generate_next_states(st.bb, st.robots, st.ans, key_config, X_wall, Y_wall);
            for (auto &ns : tmp) next_states.push_back(std::move(ns));
        }

        // 訪問マス数 + 乱数 でソートして上位だけ残す
        vector<pair<double,int>> key_idx;
        key_idx.reserve(next_states.size());
        for (int i = 0; i < (int)next_states.size(); ++i) {
            double score = next_states[i].bb.visit_count + dist(rng);
            key_idx.push_back({score, i});
        }
        sort(key_idx.begin(), key_idx.end(), [&](const auto& a, const auto& b){
            return a.first > b.first;
        });

        vector<State> trimmed;
        int keep = min(BEAM_WIDTH, (int)key_idx.size());
        trimmed.reserve(keep);
        for (int i = 0; i < keep; ++i) trimmed.push_back(std::move(next_states[key_idx[i].second]));
        current_states.swap(trimmed);
    }

    // 親機（中心 (15,15) に最も近いロボット）を選ぶ
    int main_robot_id = 0;
    {
        int best = INT_MAX;
        for (int i = 0; i < M; ++i) {
            int x = robots[i].first, y = robots[i].second;
            int dist2 = (x - 15) * (x - 15) + (y - 15) * (y - 15);
            if (dist2 < best) {
                best = dist2;
                main_robot_id = i;
            }
        }
    }

    // すべてのマスに訪れるまで（または詰まるまで）
    while (BB.visit_count < 900) {
        auto [tx, ty] = find_nearest_unvisited(BB, robots[main_robot_id].first, robots[main_robot_id].second, X_wall, Y_wall);
        char next_step = get_next_step(robots[main_robot_id].first, robots[main_robot_id].second, tx, ty, X_wall, Y_wall);
        if (next_step == 'S') break;
        int key = dir2key(next_step);
        ans_list.push_back(key);
        move_robot(BB, robots, key_config, key, X_wall, Y_wall);
    }

    // ===== 出力 =====
    // キーコンフィグ K 行（各行 M 文字をスペース区切り）
    for (int i = 0; i < K; ++i) {
        for (int j = 0; j < M; ++j) {
            if (j) cout << ' ';
            cout << key_config[i][j];
        }
        cout << '\n';
    }
    // 操作列（1 行 1 数字）
    for (int a : ans_list) cout << a << '\n';

    // 標準エラー出力にスコアを出す（デバッグ用）
    cerr << 3*N*N - ans_list.size() << endl;

    return 0;
}

File: 2025/AHC052/a04.py
import os
import sys
import time
import random
from typing import Tuple, List

MyPC = os.path.basename(__file__) != "Main.py"
if MyPC:
  from icecream import ic
  ic.disable()
else:
  def ic(*args):
    return None

ic.enable() if MyPC else None

start_time = time.time()

random.seed(1317)

# N, M, Kは固定
N = 30  # グリッドの高さ・幅
M = 10  # ロボットの数
K = 10  # キーコンフィグの数

BEAM_WIDTH = 100
TIME_LIMIT = 1.5  # 秒
LEN = 300

UDLR = ("U", "D", "L", "R")

# 各方向のキー設定
dir2key = {
  "U": 0,
  "D": 1,
  "L": 2,
  "R": 3,
  "S": 4
}

# 30x30のビットボード
# ロボットが訪れたますかどうかを管理
class BitBoard:
  def __init__(self, *args):
    # 引数があればそれをビットボードとして扱う
    if args:
      self.board = args[0]
      self.visit_count = args[1]
    # ない場合は初期化
    else:
      self.board = 1<<900  # 30x30のビットボード。C++に移植する際は64bit整数15個の配列で扱う
      self.visit_count = 0  # 訪問したマスの数。900ですべてのマスを訪問したことになる

  # ロボットが(x, y)を訪れたことを記録
  def visit(self, x: int, y: int):
    if not self.has_visited(x, y):
      self.visit_count += 1
    self.board |= 1 << (y * 30 + x)

  # (x, y)にロボットが訪れたことがあるか
  def has_visited(self, x: int, y: int) -> bool:
    return (self.board & (1 << (y * 30 + x))) != 0

# 最も少ない手順でたどり着ける未訪問マスを探す
def find_nearest_unvisited(BB: BitBoard, x: int, y: int, X_wall_list: List[List[int]], Y_wall_list: List[List[int]]) -> Tuple[int, int]:
  # BFSのためのキュー
  queue = [(x, y, 0)]  # (x, y, 手数)
  visited = set()
  visited.add((x, y))

  while queue:
    cx, cy, steps = queue.pop(0)

    # 未訪問マスを見つけたら返す
    if not BB.has_visited(cx, cy):
      return (cx, cy)

    # 移動可能な方向を探索
    for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
      nx, ny = cx + dx, cy + dy

      # 壁がないかつ範囲内であればキューに追加
      if (0 <= nx < 30) and (0 <= ny < 30) and (nx, ny) not in visited:
        if dx == -1 and X_wall_list[nx][ny] == 0:  # 上
          queue.append((nx, ny, steps + 1))
          visited.add((nx, ny))
        elif dx == 1 and X_wall_list[nx-1][ny] == 0:  # 下
          queue.append((nx, ny, steps + 1))
          visited.add((nx, ny))
        elif dy == -1 and Y_wall_list[nx][ny] == 0:  # 左
          queue.append((nx, ny, steps + 1))
          visited.add((nx, ny))
        elif dy == 1 and Y_wall_list[nx][ny-1] == 0:  # 右
          queue.append((nx, ny, steps + 1))
          visited.add((nx, ny))

  return (-1, -1)  # 見つからなかった場合

# 目的地に行くために最適な次のステップを計算
def get_next_step(BB: BitBoard, robot: Tuple[int, int], destination: Tuple[int, int], X_wall_list: List[List[int]], Y_wall_list: List[List[int]]) -> str:
  # BFSで最短経路の最初の一歩を見つける
  rx, ry = robot
  tx, ty = destination

  if rx == tx and ry == ty:
    return "S"  # 既に目的地にいる場合は停止

  queue = [(rx, ry, "")]  # (x, y, 最初の方向)
  visited = set()
  visited.add((rx, ry))

  while queue:
    cx, cy, first_direction = queue.pop(0)
    
    # 目的地に到達したら最初の方向を返す
    if cx == tx and cy == ty:
      return first_direction
    
    # 4方向に移動を試す
    directions = [(0, -1, "L"), (0, 1, "R"), (-1, 0, "U"), (1, 0, "D")]
    
    for dx, dy, direction in directions:
      nx, ny = cx + dx, cy + dy
      
      # 範囲外チェック
      if not (0 <= nx < 30 and 0 <= ny < 30):
        continue
      
      if (nx, ny) in visited:
        continue
      
      # 壁チェック
      can_move = False
      if dx == -1 and X_wall_list[nx][ny] == 0:  # 上
        can_move = True
      elif dx == 1 and X_wall_list[nx-1][ny] == 0:  # 下
        can_move = True
      elif dy == -1 and Y_wall_list[nx][ny] == 0:  # 左
        can_move = True
      elif dy == 1 and Y_wall_list[nx][ny-1] == 0:  # 右
        can_move = True

      if can_move:
        # 最初のステップの場合は方向を記録
        next_first_direction = first_direction if first_direction else direction
        queue.append((nx, ny, next_first_direction))
        visited.add((nx, ny))

  return "S"  # 経路が見つからない場合は停止

# ロボットの動きを反映
def move_robot(BB: BitBoard, robots_list: List[List[int]], key_config_list: List[List[str]], key: int, X_wall_list: List[List[int]], Y_wall_list: List[List[int]]) -> Tuple[BitBoard, List[List[int]]]:
  for i, (x, y) in enumerate(robots_list):
    direction = key_config_list[key][i]
    if direction == "U":
      nx, ny = x - 1, y
    elif direction == "D":
      nx, ny = x + 1, y
    elif direction == "L":
      nx, ny = x, y - 1
    elif direction == "R":
      nx, ny = x, y + 1
    else:
      continue

    # 移動先が範囲内かつ壁がない場合に移動
    # ic(x, y, nx, ny, direction)
    if 0 <= nx < 30 and 0 <= ny < 30:
      if direction == "U" and X_wall_list[nx][ny] == 0:
        robots_list[i] = (nx, ny)
        BB.visit(nx, ny)
      elif direction == "D" and X_wall_list[nx-1][ny] == 0:
        robots_list[i] = (nx, ny)
        BB.visit(nx, ny)
      elif direction == "L" and Y_wall_list[nx][ny] == 0:
        robots_list[i] = (nx, ny)
        BB.visit(nx, ny)
      elif direction == "R" and Y_wall_list[nx][ny-1] == 0:
        robots_list[i] = (nx, ny)
        BB.visit(nx, ny)

  return BB, robots_list

# ビームサーチの次の手を探索
def generate_next_states(BB: BitBoard, robots_list: List[List[int]], ans_list: List[str], key_config_list: List[List[str]], X_wall_list: List[List[int]], Y_wall_list: List[List[int]]) -> List[Tuple[BitBoard, List[List[int]], List[str]]]:
  next_states = []
  for i in range(4):
    new_BB = BitBoard(BB.board, BB.visit_count)
    new_robots_list = [pos[:] for pos in robots_list]

    new_BB, new_robots_list = move_robot(new_BB, new_robots_list, key_config_list, i, X_wall_list, Y_wall_list)

    next_states.append((new_BB, new_robots_list, ans_list + [i]))
  return next_states

def main():
  _, _, _ = map(int, input().split())  # N, M, Kを受け取るが使わない
  robots_list = [list(map(int, input().split())) for _ in range(M)]  # 各ロボットの座標
  Y_wall_list = [[int(c) for c in input()] for _ in range(N)]  # 左右の移動を阻害する壁の情報
  X_wall_list = [[int(c) for c in input()] for _ in range(N-1)]  # 上下の移動を阻害する壁の情報
  # ic(len(robots_list), len(X_wall_list[0]), len(Y_wall_list[0]))

  key_config_list = [["S"]*M for _ in range(K)]  # キーコンフィグの情報 U,D,L,Rで上下左右に移動、Sで停止

  key_config_list[0] = ["U"]*M  # 1手目は全ロボットを上に移動
  key_config_list[1] = ["D"]*M  # 2手目は全ロボットを下に移動
  key_config_list[2] = ["L"]*M  # 3手目は全ロボットを左に移動
  key_config_list[3] = ["R"]*M  # 4手目は全ロボットを右に移動

  BB = BitBoard()

  # ロボットの初期位置をビットボードに登録
  for x, y in robots_list:
    BB.visit(x, y)

  ans_list = []

  current_states = []
  current_states.append((BB, robots_list, []))  # (ビットボード, ロボットの位置リスト, 手順リスト)

  # 最初のLEN手は幅BEAM_WIDTHのビームサーチ
  for i in range(LEN):
    if time.time() - start_time > TIME_LIMIT:
      break
    next_states = []
    # ビームサーチの実行
    for BB, robots_list, ans_list in current_states:
      # 次の状態を生成
      tmp_next_states = generate_next_states(BB, robots_list, ans_list, key_config_list, X_wall_list, Y_wall_list)
      for state in tmp_next_states:
        next_states.append(state)
    next_states = sorted(next_states, key=lambda x: x[0].visit_count + random.random(), reverse=True)[:BEAM_WIDTH]  # 訪問マス数でソートして上位BEAM_WIDTH個を残す。ランダム性を少し加えることで多様性を確保
    current_states = next_states

  main_robot_id = 0
  # 最も中心(15, 15)に近いロボットを親機とする
  min_distance = 10**9
  for i, (x, y) in enumerate(robots_list):
    distance = (x - 15) ** 2 + (y - 15) ** 2
    if distance < min_distance:
      min_distance = distance
      main_robot_id = i


  # すべてのマスに訪れるまで繰り返す
  while BB.visit_count < 900:
    nearest_unvisited = find_nearest_unvisited(BB, robots_list[main_robot_id][0], robots_list[main_robot_id][1], X_wall_list, Y_wall_list)
    next_step = get_next_step(BB, robots_list[main_robot_id], nearest_unvisited, X_wall_list, Y_wall_list)
    # ic(robots_list[0], nearest_unvisited, next_step)
    if next_step == "S":
      break
    ans_list.append(dir2key[next_step])
    BB, robots_list = move_robot(BB, robots_list, key_config_list, dir2key[next_step], X_wall_list, Y_wall_list)  # 移動を反映
  
  # 回答を出力
  for key_config in key_config_list:
    print(*key_config)
  for ans in ans_list:
    print(ans)

  # スコアを標準エラー出力に出力
  if MyPC:
    total_score = 3 * N**2 - len(ans_list)
    print(total_score, file=sys.stderr)

if __name__ == "__main__":
  main()